### PR TITLE
Document Style Guide

### PR DESCRIPTION
**Pull request by UnloadingGnat**
#15 

## Overview

Ported Huff style guide from [notion](https://coherent-wolfberry-04a.notion.site/huff-lib-Styling-9c2ceaa795934c38be0b2f2d10042fe9). 

## Checklist

- [x] Ported Style Guide
- [x] Fixed Expired Discord links
- [x] Fix Alignment issues
- [x] Use __FUNC_SIG()

### CODE REVIEW

File: src/style-guide/overview/README.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+jump
+
+// ....
+
+operation0:
+	  0x01     // [b]
+	  0x02     // [a, b]
+	  add      // [sum]
+operation1:
+	  0x02     // [b, sum]
```
Comment (abigger87) [inline src/style-guide/overview/README.md:185]: align

File: src/style-guide/overview/README.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+}
+
+#define macro DO_ACTION() = takes (0) returns (0) {
+		0x45    // [value]
+    0x00    // [offset, value]
+	  mstore  // []
+    0x20    // [size]
+    0x00    // [offset, size]
+    return
+}
```
Comment (abigger87) [inline src/style-guide/overview/README.md:248]: ibid

File: src/style-guide/overview/README.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+    sstore       // []
+}
+
+#define macro DO_ACTION() = takes (0) returns (2) {
+		0x45    // [value]
+    0x00    // [offset, value]
+	  mstore  // []
+    0x20    // [size]
+    0x00    // [offset, size]
+}
```
Comment (abigger87) [inline src/style-guide/overview/README.md:303]: align

File: src/style-guide/overview/README.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+		other_action:
+	    OTHER_ACTION()
+			finish jump
+
+    finish:
+	    // stack: [offset, size]
+			UNLOCK()
+			return
+}
```
Comment (abigger87) [inline src/style-guide/overview/README.md:328]: align + `__FUNC_SIG()`

**Comment by UnloadingGnat** [comment]
working on fixes/alignment now.

File: src/contribute/overview/README.md [hunk 5]
```diff
@@ -14,7 +14,7 @@ All contributions are welcome! We want to make contributing to this project as e
 - Becoming a maintainer
 
 
-Contributing to Huff Language is not just limited to writing code. It can mean a wide array of things, such as participating in discussions in our [discord server](https://discord.gg/h8pkspwx) or reporting issues.
+Contributing to Huff Language is not just limited to writing code. It can mean a wide array of things, such as participating in discussions in our [discord server](https://discord.gg/PfH8TZaYxQ) or reporting issues.
```
[the remark this answers is not in the document]
Comment (UnloadingGnat) [reply src/contribute/overview/README.md:17]: Ok, I see. There are a bunch of those discord links in the docs, including the footer. I'll fix them all in this PR.

File: src/style-guide/overview/README.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+#define macro ADD() = takes (0) returns (0) {}
+
+#define macro MAIN() = takes (0) returns (0) {
+		0x00 calldataload 0xE0 shr
+    dup1 ADD_SEL eq add_func jumpi
+
+		add_func:
+			ADD()
+}
```
[the remark this answers is not in the document]
Comment (UnloadingGnat) [reply src/style-guide/overview/README.md:150]: Is this the right way of using __FUNC_SIG(), or do I need to remove the macro ADD()?

```solidity
#define function add(uint256, uint256) view returns (uint256)

#define macro ADD() = takes (0) returns (0) {}

#define macro MAIN() = takes (0) returns (0) {
    // Grab the function selector from the calldata
    0x00 calldataload 0xE0 shr                 // [selector]
    dup1 __FUNC_SIG(add) eq add_func jumpi     // [selector]

    // Revert if no functions match
    0x00 0x00 revert

    add_func:
        ADD()
}
```

File: src/style-guide/overview/README.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+#define macro MAIN() = takes (0) returns (0) {
+		0x00 calldataload 0xE0 shr
+    dup1 DO_ACTION eq do_action jumpi
+
+		do_action:
+    NON_REENTRANT()
+    DO_ACTION()
+		UNLOCK()
+}
```
[the remark this answers is not in the document]
Comment (UnloadingGnat) [reply src/style-guide/overview/README.md:258]:
```solidity
#define function doAction() view returns (uint256)

#define constant LOCK_SLOT = FREE_STORAGE_POINTER()

#define macro NON_REENTRANT() = takes (0) returns (0) {
    [LOCK_SLOT]  // [lock_slot]
    sload        // [lock]
    iszero       // [is_unlocked]
    unlocked     // [unlocked_jumpdest]
    jumpi        // []
    0x00         // [size]
    0x00         // [offset, size]
    revert       // []
    unlocked:    // []
    0x01         // [lock_value]
    [LOCK_SLOT]  // [lock_slot, lock_value]
    sstore       // []
}

#define macro UNLOCK() = takes (0) returns (0) {
    0x01         // [lock_value]
    [LOCK_SLOT]  // [lock_slot, lock_value]
    sstore       // []
}

#define macro DO_ACTION() = takes (0) returns (0) {
    0x45    // [value]
    0x00    // [offset, value]
    mstore  // []
    0x20    // [size]
    0x00    // [offset, size]
    return
}

#define macro MAIN() = takes (0) returns (0) {
    0x00 calldataload 0xE0 shr                       // [selector]
    dup1 __FUNC_SIG(doAction) eq do_action jumpi     // [selector]

    // Revert if no functions match
    0x00 0x00 revert

    do_action:
        NON_REENTRANT()
        DO_ACTION()
        UNLOCK()
}
```

File: src/style-guide/overview/README.md (hunk 4 above)
Comment (UnloadingGnat) [reply src/style-guide/overview/README.md:328]:
```solidity
#define function doAction() view returns (uint256)
#define function otherAction() view returns (uint256)

#define constant LOCK_SLOT = FREE_STORAGE_POINTER()

#define macro NON_REENTRANT() = takes (0) returns (0) {
    [LOCK_SLOT]  // [lock_slot]
    sload        // [lock]
    iszero       // [is_unlocked]
    unlocked     // [unlocked_jumpdest]
    jumpi        // []
    0x00         // [size]
    0x00         // [offset, size]
    revert       // []
    unlocked:    // []
    0x01         // [lock_value]
    [LOCK_SLOT]  // [lock_slot, lock_value]
    sstore       // []
}

#define macro UNLOCK() = takes (0) returns (0) {
    0x00         // [lock_value]
    [LOCK_SLOT]  // [lock_slot, lock_value]
    sstore       // []
}

#define macro DO_ACTION() = takes (0) returns (2) {
    0x45    // [value]
    0x00    // [offset, value]
    mstore  // []
    0x20    // [size]
    0x00    // [offset, size]
}

#define macro OTHER_ACTION() = takes (0) returns (2) {
    0x00    // [size]
    0x00    // [offset, size]
}

#define macro MAIN() = takes (0) returns (0) {
    // Grab the function selector from the calldata
    0x00 calldataload 0xE0 shr                          // [selector]
    dup1 __FUNC_SIG(doAction) eq do_action jumpi        // [selector]
    dup1 __FUNC_SIG(otherAction) eq other_action jumpi  // [selector]

    // Revert if no functions match
    0x00 0x00 revert


    do_action:
        NON_REENTRANT()
        DO_ACTION()
        finish jump

    other_action:
        OTHER_ACTION()
        finish jump

    finish:
        // stack: [offset, size]
        UNLOCK()
        return
}
```

File: src/style-guide/overview/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (abigger87) [reply src/style-guide/overview/README.md:150]: yup!

File: src/style-guide/overview/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (abigger87) [reply src/style-guide/overview/README.md:258]: Nice!

File: src/style-guide/overview/README.md (hunk 4 above)
Comment (abigger87) [reply src/style-guide/overview/README.md:328]: Beautiful!

**Comment by UnloadingGnat** [comment]
- alignment fixed
- replaced Labels with __FUNC_SIG()

Merge Conflict resolved, had to do with resources linking to /resources/overview in config.js

**Review by abigger87 — APPROVED** [review]
LGTM

**Pull request merged**